server/src/main/java/org/apache/cassandra/sidecar/coordination/InnerDcTokenAdjacentPeerProvider.java

```diff
@@ -0,0 +1,270 @@
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.google.common.base.Preconditions;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datastax.driver.core.DataType;
+import com.datastax.driver.core.Host;
+import com.datastax.driver.core.KeyspaceMetadata;
+import com.datastax.driver.core.Metadata;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import org.apache.cassandra.sidecar.cluster.InstancesMetadata;
+import org.apache.cassandra.sidecar.cluster.instance.InstanceMetadata;
+import org.apache.cassandra.sidecar.common.client.SidecarInstance;
+import org.apache.cassandra.sidecar.common.client.SidecarInstanceImpl;
+import org.apache.cassandra.sidecar.common.server.cluster.locator.Token;
+import org.apache.cassandra.sidecar.common.server.cluster.locator.TokenRange;
+import org.apache.cassandra.sidecar.common.server.dns.DnsResolver;
+import org.apache.cassandra.sidecar.config.ServiceConfiguration;
+
+
+import static org.apache.cassandra.sidecar.config.yaml.CassandraInputValidationConfigurationImpl.DEFAULT_FORBIDDEN_KEYSPACES;
+
+
+/**
+ * Return Sidecar(s) adjacent to current Sidecar in the token ring within the same datacenter.
+ */
+@Singleton
+public class InnerDcTokenAdjacentPeerProvider implements SidecarPeerProvider
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(InnerDcTokenAdjacentPeerProvider.class);
+
+    protected final InstancesMetadata instancesMetadata;
+    private final CassandraClientTokenRingProvider cassandraClientTokenRingProvider;
+    private final ServiceConfiguration serviceConfiguration;
+    private final DnsResolver dnsResolver;
+
+    @Inject
+    public InnerDcTokenAdjacentPeerProvider(InstancesMetadata instancesMetadata,
+                                            CassandraClientTokenRingProvider cassandraClientTokenRingProvider,
+                                            ServiceConfiguration serviceConfiguration,
+                                            DnsResolver dnsResolver)
+    {
+        this.instancesMetadata = instancesMetadata;
+        this.cassandraClientTokenRingProvider = cassandraClientTokenRingProvider;
+        this.serviceConfiguration = serviceConfiguration;
+        this.dnsResolver = dnsResolver;
+    }
+
+    public Set<SidecarInstance> get()
+    {
+        Map<Integer, InstanceMetadata> localInstances = instancesMetadata
+                                                        .instances()
+                                                        .stream()
+                                                        .collect(Collectors.toMap(InstanceMetadata::id, Function.identity()));
+
+        if (localInstances.isEmpty())
+        {
+            LOGGER.debug("No local instances found");
+            return Set.of();
+        }
```

Comment:
With `InstanceMetadataFetcher`, this `if` statement can be removed too. It is covered by the suggestion for line#99